File: destiny-engine/tests/backtest.rs
use destiny_engine::prelude::*;

struct BacktestStrategy {
    symbol: String,
    is_buy: Arc<Mutex<bool>>,
}

impl BacktestStrategy {
    pub fn new() -> Self {
        Self {
            symbol: "ETHUSDT".to_string(),
            is_buy: Arc::new(Mutex::new(false)),
        }
    }
}

#[async_trait]
#[allow(unused_variables)]
impl Strategy for BacktestStrategy {
    async fn on_init(&self, engine: Arc<dyn Engine>) -> Result<()> {
        tracing::info!("{} on_init", engine.time());
        engine.symbol_init(&self.symbol)?;
        Ok(())
    }

    async fn on_minutely(&self, engine: Arc<dyn Engine>) -> Result<()> {
        if !*self.is_buy.lock() {
            engine
                .long_limit_open(&self.symbol, Decimal::new(1.0), Decimal::new(200.0))
                .await?;
            *self.is_buy.lock() = true;
        }
        let time = engine.time().str_ymd_hm();
        let price_mark = engine.price_mark(&self.symbol);
        let cash_available = engine.cash_available();
        info!("{time} 标记价({price_mark:.2}),可用资金({cash_available:.2})");
        Ok(())
    }
}

#[tokio::test(flavor = "multi_thread")]
async fn test_backtest() -> Result<()> {
    if bool::has_github_action() {
        return Ok(());
    }

    let log_collector = LogConfigBuilder::default()
        .level(LogLevel::INFO)
        .save_file(false)
        .targets(vec!["backtest".to_string()])
        .build()?
        .init_log()
        .await?;

    let config = BacktestConfigBuilder::default()
        .begin("2020".to_date()?)
        .end("2024".to_date()?)
        .build()?;

    Backtest::run(config, Arc::new(BacktestStrategy::new())).await?;

    log_collector.done().await?;

    Ok(())
}

File: destiny-python/python/destiny/tests/backtest.py
from destiny import *


class BacktestStrategy(Strategy):
    symbol: str
    isbuy: bool

    def __init__(self):
        self.symbol = "ETHUSDT"
        self.isbuy = False

    def on_init(self, api: API):
        api.init_symbol(self.symbol)

    def on_start(self, api: API):
        pass

    def on_stop(self, api: API):
        pass

    def on_daily(self, api: API):
        if not self.isbuy:
            api.long_limit_open(self.symbol, Decimal(0.1), Decimal(3000))
            self.isbuy = True
        time = api.time()
        price_mark = api.price_mark(self.symbol)
        cash_available = api.cash_available()
        margin = api.margin()
        long_size = api.long_size(self.symbol)
        info(
            f"{time} 标记价({price_mark:.2f}),可用资金({cash_available:.4f}),保证金({margin:.4f}),多仓({long_size:.4f})"
        )

    def on_hourly(self, api: API):
        pass

    def on_minutely(self, api: API):
        pass

    def on_kline(self, api: API, kline: Kline):
        pass

    def on_order(self, api: API, order: Order):
        pass

    def on_position(self, api: API, position: Position):
        pass


init_log(show_std=True, save_file=True)
run_backtest(BacktestStrategy(), "2023", "2024")
